### ISSUE TITLE
Changing Node IP orphans CEPs

### ISSUE THREAD

**Issue by bprashanth** [cilium/cilium#26279]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

Repro: 

1. In an environment where cilium serializes EPs into tmpfs (eg `/var/run` - which is currently the case on some cloud providers)
2. When node IP changes (typically on reboot or suspend/resume in the cloud)
3. CEPs are orphaned by the restarted cilium 
4. This results in ip cache corruption and a disruption of pod to pod traffic 

Root cause: 

The [endpointsynchronizer](https://github.com/cilium/cilium/blame/main/pkg/k8s/watchers/endpointsynchronizer.go) is responsible for updating the status of the CEP on cilium restart. 

When the cilium restart is because of a node restart which results in a node IP change, the endpoint synchronizer gets into a soft deadlock. 

* The [UID check](https://github.com/cilium/cilium/blob/main/pkg/k8s/watchers/endpointsynchronizer.go#L344) fails because it compares the CEP's UID with the previously serialized CEP's UID. Since the CEP was serialized into tmpfs, this state is lost. 
* The [Node IP check](https://github.com/cilium/cilium/blob/main/pkg/k8s/watchers/endpointsynchronizer.go#L357) fails because the node IP has changed. 

Instead of comparing the CEP node ip with the node's ip, [this](https://github.com/cilium/cilium/pull/26281) poc looks up the pod for the CEP and uses its host IP. The latter is updated by the kubelet and hence avoids the deadlock.

Opening this issue for discussion on what the right fix should be. 
@tommyp1ckles 

### Cilium Version

Tested on 
```
Client: 1.12.6 098dd3b4ba 2023-05-15T14:15:53-07:00 go version go1.20.1 linux/amd64
Daemon: 1.12.6 098dd3b4ba 2023-05-15T14:15:53-07:00 go version go1.20.1 linux/amd64
```
But should repro on head based on code inspection 

### Kernel Version

5.15.65+

### Kubernetes Version

```
Client Version: version.Info{Major:"1", Minor:"27", GitVersion:"v1.27.3", GitCommit:"25b4e43193bcda6c7328a6d147b1fb73a33f1598", GitTreeState:"clean", BuildDate:"2023-06-14T09:53:42Z", GoVersion:"go1.20.5", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v5.0.1
Server Version: version.Info{Major:"1", Minor:"26", GitVersion:"v1.26.3-gke.1000", GitCommit:"f14ee8d868842856ce474ba2ebadece18a6ca6cf", GitTreeState:"clean", BuildDate:"2023-03-30T09:31:19Z", GoVersion:"go1.19.7 X:boringcrypto", Compiler:"gc", Platform:"linux/amd64"}
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by christarazi** [comment]
Is [this](https://github.com/cilium/cilium/issues/25094) the same issue you observe?

**Comment by tommyp1ckles** [comment]
> Is [this](https://github.com/cilium/cilium/issues/25094) the same issue you observe? 

I think its sort of a combination of the two ownership issues I've fixed. Normally in this case the uid check is what allows the agent to take ownership of a CEP with a different node Ip.

What I didn't consider was a combination of the missing state and node Ip change. 

As well the init CEP gc won't clean this up because it's not a "local" endpoint which it ignores.

**Comment by aojea** [comment]
> As well the init CEP gc won't clean this up because it's not a "local" endpoint which it ignores.

so, a same node name with different nodeIP should trigger a cache recalculation?

**Comment by tommyp1ckles** [comment]
> code freeze
correct

Thanks for the PR 🎉  I think the logic makes sense, I would be a little concerned with hard-wiring API server get requests into that logic for every Pod, this might potentially cause issues in the case of a lot of nodes restarting and needing to hammer the API server with requests.

I think a better approach would be to ensure that EPS doesn't start until the k8s Pods cache is synced then use that to get Pod Node IP references.

Presumably the Pod resources themselves have to be updated with the new Node IP, do we know if that will always occur prior to Agent starting?

**Comment by tommyp1ckles** [comment]
Using node names might suffer from the same issue, because upon restart Kubelet can say, reregister itself with a new node name (i.e. possibly derived from the IP?)

**Comment by aojea** [comment]
> I think a better approach would be to ensure that EPS doesn't start until the k8s Pods cache is synced then use that to get Pod Node IP references.

if you have an informer use the pod cache, I assume you are already waiting for informer cache to be in sync

> Using node names might suffer from the same issue

kubernetes objects have UIDs that univocally identifies an object, however, I recognize I'm a bit lost in this conversation and I don't think that is the problem you are trying to solve

**Comment by bprashanth** [comment]
```
Presumably the Pod resources themselves have to be updated with the new Node IP, do we know if that will always occur prior to Agent starting?
```
No it wont happen before the Agent starts. Because the kubelet updates it in paralle.
But the agent should keep retrying this check till it passes. And eventually it will go through, once the kubelet has updated it. I'll confirm this. 

I will also look at how complicated sharing the `k8s/watcher/pod` informer cache is going to be. 

```
Using node names might suffer from the same issue, because upon restart Kubelet can say, reregister itself with a new node name (i.e. possibly derived from the IP?)
```
You're right, the kubelet has a hostname-override flag to overwrite the machines hostname. 
However if this changes, the old pod will get deleted because the kubelet will disown it, and a new pod will get created, automatically triggering the right path. 

Anyway for now I think I'll evaluate the feasibility of using pod.HostIP and informer cache. There might be more edge cases that need thinking through with the hostname approach. 

```
kubernetes objects have UIDs that univocally identifies an object, however, I recognize I'm a bit lost in this conversation and I don't think that is the problem you are trying to solve
```

The point of the ip/hostname check is to make sure this node (i.e the node cilium is running on, as computed using node and CN ip or hostname) and the node the pod is running on (the pod for the CEP this cilium is trying to manage) are the same.

I'm wondering if we can use pod uid for this instead.     
* If the pod is NOT deleted/recreated, its UID will be the same but it's hostIP could be different. As long as this UID matches the CEP's ownerRef, we should be OK owning it and updating the hostIP to match. 
* If the pod is deleted/recreated (even with the same ns/name), its UID will change. 
  - If it lands on a different node, assume another cilium is owning it
  - If it lands on the same node, we would need to delete the old one and recreate a new one 

Is this preferable long term? 
There are some complications here. Specifically, what happens in the last scenario today - is the owner ref just wrong? I need to think about this, and it seems (at least mentally) more complicated than the initial ip check. 

So in summary    
1. I'll look at sharing the pod cache for the ip check
2. LMK if either of the other approaches seem preferable

**Comment by bprashanth** [comment]
1. I've updated the pr to use pod cache. 

2. I investigated using UID instead of hostIP. The cilium-operator + agent don't seem to handle the case mentioned above. But I wasn't able to trigger this situation either. I suspect when the same pod ns/name is recreated on the same node, it could result in a mismatch in pod UID and CEP `ownerReference.UID`. So I don't want to rely on this.

3. I think we need to keep syncing this ownership/dis-ownership forever. That will force eventual consistency. I will look at a follow up pr for this. 

@tommyp1ckles wondering if we can get the fix in now while i continue trying to improve it (by following up on 3, investigating 2 etc) at a lower priority? I think it's strictly better now than it was previously.

**Comment by tommyp1ckles** [comment]
> But the agent should keep retrying this check till it passes. And eventually it will go through, once the kubelet has updated it. I'll confirm this.

> I think we need to keep syncing this ownership/dis-ownership forever. That will force eventual consistency. I will look at a follow up pr for this.

Good point - that'd be great, perhaps add a comment explaining this.

Yeah I think the key thing here is being sure that this will eventually become consistent. This seems like an improvement on relying on the CEP Node IP since that fails in the situation you described.

I think your changes make sense given the approach we've taken here.  I do wonder if there's a better way we can ensure correct ownership - every fix we introduce tends to miss some edge case 😆

**Comment by bprashanth** [comment]
You're probably right, there is a better overall fix. I found myself going down that path, and then realizing that we're better off putting in the most scoped, surgical fix possible for the immediate term and working out the edge cases slowly. I can create another bug once this one's closed and look at the endpoint syncer in more detail.

**Comment by christarazi** [comment]
@bprashanth Have you followed up with a new issue?
